packages/util-dns/src/NodeDnsLookupHostResolver.ts

```diff
@@ -20,20 +23,144 @@ const DNS_LOOKUP_OPTIONS = {
   verbatim: true,
 } as const;
 
+/**
+ * Node.js dns.lookup() function type used in {@link NodeDnsLookupHostResolver}
+ * @internal
+ */
+interface NodeDnsLookupFn {
+  (hostname: string, options: LookupAllOptions): Promise<LookupAddress[]>;
+}
+
+/**
+ * Configuration options for {@link NodeDnsLookupHostResolver}
+ * @internal
+ */
+export interface NodeDnsLookupHostResolverConfig {
+  /**
+   * Cache for holding address entries
+   * @internal
+   */
+  cache?: DnsCache;
+  /**
+   * TTL offset in MS when generating expiration TTLs
+   * @internal
+   */
+  ttlMs?: number;
+  /**
+   * Pluggable Node.js dns.lookup() function
+   * @internal
+   */
+  nodeDnsLookup?: NodeDnsLookupFn;
+}
+
 /**
  * {@link HostResolver} implementation that uses the Node.js dns.lookup() API.
- * TODO(dns): implement cache
  */
 export class NodeDnsLookupHostResolver implements IHostResolver {
+  /**
+   * Default TTL in ms
+   * @internal
+   */
+  private static DEFAULT_TTL_MS = 30_000;
+
+  /**
+   * Provider that returns a default {@link DnsCache}
+   * @returns DnsCache implementation
+   * @internal
+   */
+  private static DEFAULT_CACHE_PROVIDER = () => new HostEntryTable();
+
+  /**
+   * Default Node.js dns.lookup() function
+   * @internal
+   */
+  private static DEFAULT_NODE_DNS_LOOKUP: NodeDnsLookupFn = dnsPromises.lookup;
+
+  /**
+   * TTL in ms
+   * @internal
+   */
+  private ttlMs;
+
+  /**
+   * {@link DnsCache} which maps {@link HostAddress.hostName} to {@link DnsCacheEntry}
+   * @internal
+   */
+  private cache: DnsCache;
+
+  /**
+   * Node.js dns.lookup() API implementation
+   * @internal
+   */
+  private nodeDnsLookup: NodeDnsLookupFn;
+
+  constructor(config?: NodeDnsLookupHostResolverConfig) {
+    this.ttlMs = config?.ttlMs !== undefined ? config.ttlMs : NodeDnsLookupHostResolver.DEFAULT_TTL_MS;
+    this.cache = config?.cache !== undefined ? config.cache : NodeDnsLookupHostResolver.DEFAULT_CACHE_PROVIDER();
+    this.nodeDnsLookup =
+      config?.nodeDnsLookup !== undefined ? config.nodeDnsLookup : NodeDnsLookupHostResolver.DEFAULT_NODE_DNS_LOOKUP;
+  }
```

Comment:
this can be destructured with defaults:

```ts
public constructor({
  ttlMs = NodeDnsLookupHostResolver.DEFAULT_TTL_MS,
  cache = NodeDnsLookupHostResolver.DEFAULT_CACHE_PROVIDER(),
  nodeDnsLookup = NodeDnsLookupHostResolver.DEFAULT_NODE_DNS_LOOKUP,
}: NodeDnsLookupHostResolverConfig = {}) {
  this.ttlMs = ttlMs;
  this.cache = cache;
  this.nodeDnsLookup = nodeDnsLookup;
}
```

Also, NodeDnsLookupHostResolver.DEFAULT_CACHE_PROVIDER() is a static method? Probably shouldn't be screaming snake case.

Reply:
Resolved in https://github.com/aws/aws-sdk-js-v3/pull/4433/commits/ed00d144ed0799f846797df9c989a5e03ddbb32d